src/features/broker-protection/captcha.js

```diff
@@ -0,0 +1,106 @@
+    const captcha = getElement(captchaDiv, '[src^="https://www.google.com/recaptcha"]') ||
+        getElement(captchaDiv, '[src^="https://newassets.hcaptcha.com/captcha"')
+
+    // ensure we have the elements
+    if (!captcha) return new ErrorResponse({ actionID: action.id, message: 'could not find captcha' })
+    if (!('src' in captcha)) return new ErrorResponse({ actionID: action.id, message: 'missing src attribute' })
+
+    const captchaUrl = String(captcha.src)
+    let captchaType
+    let siteKey
+
+    if (captchaUrl.includes('recaptcha/api2')) {
+        captchaType = 'recaptcha2'
+        siteKey = new URL(captchaUrl).searchParams.get('k')
+    } else if (captchaUrl.includes('recaptcha/enterprise')) {
+        captchaType = 'recaptchaEnterprise'
+        siteKey = new URL(captchaUrl).searchParams.get('k')
+    } else if (captchaUrl.includes('hcaptcha.com/captcha/v1')) {
+        captchaType = 'hcaptcha'
+        // hcaptcha sitekey may be in either
+        if (captcha instanceof Element) {
+            siteKey = captcha.getAttribute('data-sitekey')
+        }
+        if (!captcha) {
+            try {
+                // `new URL(...)` can throw, so it's valid to wrap this in try/catch
+                siteKey = new URL(captchaUrl).searchParams.get('sitekey')
+            } catch (e) {
+                console.warn('error parsing captchaUrl', captchaUrl)
+            }
+        }
+    }
+
+    if (!captchaType) {
+        return new ErrorResponse({ actionID: action.id, message: 'Could not extract captchaType.' })
+    }
+    if (!siteKey) {
+        return new ErrorResponse({ actionID: action.id, message: 'Could not extract siteKey.' })
+    }
+
+    // Remove query params (which may include PII)
+    const pageUrlWithoutParams = pageUrl?.split('?')[0]
+
+    const responseData = {
+        siteKey,
+        url: pageUrlWithoutParams,
+        type: captchaType
+    }
+
+    return new SuccessResponse({ actionID: action.id, actionType: action.actionType, response: responseData })
+}
+
+/**
+* Takes the solved captcha token and injects it into the page to solve the captcha
+*
+* @param action
+* @return {import('./types.js').ActionResponse}
+*/
+export function solveCaptcha (action, token) {
+    const selectors = ['h-captcha-response', 'g-recaptcha-response']
+    let solved = false
+
+    for (const selector of selectors) {
+        const match = document.getElementsByName(selector)[0]
+        if (match) {
+            match.innerHTML = token
+            solved = true
+            break
+        }
+    }
+
+    if (solved) {
+        const json = JSON.stringify({ token })
+
+        const javascript = `;(function(args) {
+            ${captchaCallback.toString()};
+            captchaCallback(args);
+        })(${json});`
+
+        return new SuccessResponse({
+            actionID: action.id,
+            actionType: action.actionType,
+            response: { callback: { eval: javascript } }
+        })
+    }
```

Comment:
This is where we send some JS back to native to allow them to execute it in the 'page world'. We control all of the input here - it's just `captchaCallback` stringified, with `{token: "abc"}` on the end.